eZ/Publish/Core/Persistence/Legacy/Content/Type/Gateway/DoctrineDatabase.php

```diff
@@ -1540,4 +1540,57 @@ public function removeFieldDefinitionTranslation(
 
         $deleteQuery->execute();
     }
+
+    /**
+     * Removes types created or modified by the user.
+     */
+    public function removeByUserAndVersion(int $userId, int $version): void
+    {
+        $this->connection->beginTransaction();
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb->delete('ezcontentclass')
+            ->where('creator_id = :user or modifier_id = :user')
+            ->andWhere('version = :version')
+            ->setParameter('user', $userId, ParameterType::INTEGER)
+            ->setParameter('version', $version, ParameterType::INTEGER);
+
+        $qb->execute();
+
+        $this->cleanupAssociations();
+
+        $this->connection->commit();
+    }
+
+    private function cleanupAssociations()
+    {
+        $this->cleanupClassAttributeTabel();
+        $this->cleanupClassAttributeMLTabel();
+        $this->cleanupClassGroupTabel();
+        $this->cleanupClassNameTabel();
+    }
+
+    private function cleanupClassAttributeTabel()
+    {
+        $sql = 'delete from ezcontentclass_attribute where not exists (select 1 from ezcontentclass where id = ezcontentclass_attribute.contentclass_id and version = ezcontentclass_attribute.version)';
```

Comment:
Can we format those strings, like `DELETE FROM ezcontentclass_attribute WHERE ...` They will be more readable